solutions/usaco-571.mdx

```diff
@@ -2,55 +2,167 @@
 id: usaco-571
 source: USACO Silver 2015 December
 title: High Card Wins
-author: Óscar Garries
+author: Óscar Garries, Ryan Chou
 ---
 
-[Official Analysis](http://www.usaco.org/current/data/sol_highcard_silver_dec15.html)
+[Official Analysis (Java)](http://www.usaco.org/current/data/sol_highcard_silver_dec15.html)
 
-<LanguageSection>
-<CPPSection>
+Intuitively, we always want Bessie to win with the smallest card possible. We can ensure this minimum by storing the available cards for both Bessie and Elsie. If Elsie's current card is higher than Bessie's, we can move to the Bessie's next higher card.
+
+You can see the algorithm in action with the test set below:
+
+<img src = "/usaco-571-anim/highcard_algo.gif"/>
 
 ## Implementation
 
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+
+<CPPSection>
+
 ```cpp
-#include <bits/stdc++.h>
+#include <iostream>
+#include <vector>
+#include <unordered_set>
+#include <algorithm>
+#include <cstdio>
 
 using namespace std;
 
-int main () {
+int main() {
 	freopen("highcard.in", "r", stdin);
 	freopen("highcard.out", "w", stdout);
 
+	unordered_set<int> elsieHas;
```

Comment:
any reason you use unordered_set as opposed to just set? IIRC unordered can be slow

also since the cards are from 1 to 2 * n, you can (and it might be more intuitive to) use a boolean array

Reply:
also the java solution uses one so change for consistency?

Reply:
sure!